encoding/series_iterator.go

```diff
@@ -0,0 +1,225 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package encoding
+
+import (
+	"container/heap"
+	"errors"
+	"time"
+
+	"github.com/m3db/m3db/interfaces/m3db"
+	xtime "github.com/m3db/m3db/x/time"
+)
+
+var (
+	timeZero              time.Time
+	errOutOfOrderIterator = errors.New("series values are out of order from mixed reader")
+)
+
+type seriesIterator struct {
+	id            string
+	start         time.Time
+	end           time.Time
+	iters         []m3db.Iterator
+	currIters     IteratorHeap
+	unclosedIters []m3db.Iterator
+	err           error
+	firstNext     bool
+	closed        bool
+	pool          m3db.SeriesIteratorPool
+}
+
+// NewSeriesIterator creates a new series iterator
+func NewSeriesIterator(
+	id string,
+	startInclusive, endExclusive time.Time,
+	replicas []m3db.Iterator,
+	pool m3db.SeriesIteratorPool,
+) m3db.SeriesIterator {
+	it := &seriesIterator{pool: pool}
+	it.Reset(id, startInclusive, endExclusive, replicas)
+	return it
+}
+
+func (it *seriesIterator) ID() string {
+	return it.id
+}
+
+func (it *seriesIterator) Start() time.Time {
+	return it.start
+}
+
+func (it *seriesIterator) End() time.Time {
+	return it.end
+}
+
+func (it *seriesIterator) Next() bool {
+	if it.firstNext {
+		it.firstNext = false
+		return it.hasNext()
+	}
+	if !it.hasNext() {
+		return false
+	}
+	it.moveToNext(true)
+	return it.hasNext()
+}
+
+func (it *seriesIterator) Current() (m3db.Datapoint, xtime.Unit, m3db.Annotation) {
+	return it.currIters[0].Current()
+}
+
+func (it *seriesIterator) Err() error {
+	return it.err
+}
+
+func (it *seriesIterator) Close() {
+	if it.isClosed() {
+		return
+	}
+	it.closed = true
+	for _, iter := range it.iters {
+		iter.Close()
+	}
+	it.iters = it.iters[:0]
+	if it.pool != nil {
+		it.pool.Put(it)
+	}
+}
+
+func (it *seriesIterator) Reset(id string, startInclusive, endExclusive time.Time, replicas []m3db.Iterator) {
+	it.id = id
+	it.start = startInclusive
+	it.end = endExclusive
+	it.iters = it.iters[:0]
+	it.currIters = it.currIters[:0]
+	it.unclosedIters = it.unclosedIters[:0]
+	it.err = nil
+	it.firstNext = true
+	it.closed = false
+	for _, replica := range replicas {
+		// Replica can be nil if replica failed to respond
+		if replica == nil {
+			continue
+		}
+		it.iters = append(it.iters, replica)
+	}
+	heap.Init(&it.currIters)
+	it.moveToNext(false)
+}
+
+func (it *seriesIterator) hasError() bool {
+	return it.err != nil
+}
+
+func (it *seriesIterator) isClosed() bool {
+	return it.closed
+}
+
+func (it *seriesIterator) hasMore() bool {
+	return len(it.iters) > 0
+}
+
+func (it *seriesIterator) hasNext() bool {
+	return !it.hasError() && !it.isClosed() && it.hasMore()
+}
+
+func (it *seriesIterator) moveIterToValidNext(iter m3db.Iterator) bool {
+	dp, _, _ := iter.Current()
+	prevT := dp.Timestamp
```

Comment:
Not quite, because I do a `t.Before(prevT)` this allows for the situation you've just described.  So it can be the same or in the future, just not before.